pkg/promtail/client/client.go

```diff
@@ -155,8 +158,18 @@ func (c *Client) sendBatch(batch map[model.Fingerprint]*logproto.Stream) {
 	var status int
 	for backoff.Ongoing() {
 		start := time.Now()
-		status, err = c.send(ctx, buf)
+		timedCtx, ctxCancel := context.WithTimeout(ctx, c.cfg.Timeout)
+		status, err = c.send(timedCtx, buf)
 		requestDuration.WithLabelValues(strconv.Itoa(status)).Observe(time.Since(start).Seconds())
+		ctxErr := timedCtx.Err()
+		if ctxErr != nil {
+			switch ctxErr {
+			case context.DeadlineExceeded:
+				err = errors.Wrap(err, "Timeout waiting for server")
+			}
+		}
+		ctxCancel()
```

Comment:
We probably want to move the `context.WithTimeout` into `client.send` and then defer the `ctxCancel` call.  And just call it `cancel`.